src/runtime/tree.c

```diff
@@ -132,13 +141,25 @@ void ts_tree_retain(TSTree *self) {
 }
 
 void ts_tree_release(TSTree *self) {
+  if (!self)
+    return;
+
+recur:
   assert(self->ref_count > 0);
   self->ref_count--;
+
   if (self->ref_count == 0) {
-    for (size_t i = 0; i < self->child_count; i++)
-      ts_tree_release(self->children[i]);
-    if (self->child_count > 0)
+    if (self->child_count > 0) {
+      for (size_t i = 0; i < self->child_count - 1; i++)
+        ts_tree_release(self->children[i]);
+      TSTree *last_child = self->children[self->child_count - 1];
       ts_free(self->children);
+      ts_free(self);
+
+      self = last_child;
+      goto recur;
+    }
+
```

Comment:
This manual tail call optimization, and the one above in `ts_tree_assign_parents`, make the stack depth a lot more predictable when parsing large files:

before:
![before-tail-call-opt](https://cloud.githubusercontent.com/assets/326587/12829573/c00d70ca-cb3f-11e5-9c3a-036df57c799f.png)

after:
![after-tail-call-opt](https://cloud.githubusercontent.com/assets/326587/12829583/c5ff4cd8-cb3f-11e5-8ee4-ebd17ec3d4f8.png)

For reasonable inputs, the stack depth is now effectively constant, since there is an upper bound on the depth of real programs' concrete syntax trees. It is still possible for a malicious user to create a pathological file consisting only of thousands of nested parentheses, or something like that. In that case, the stack could overflow. To guard agains that, I think it would be sufficient to track the depth explicitly, and just error out if an unreasonably deep tree is encountered. I'll leave that to a later PR though.